Display New Version information from changelog at bottom of Info Page

kdk1616 opened the pull request:
* Parses Markdown from CHANGELOG.md and displays at bottom of info page

kdk1616 committed:
Added Changelog to Info Tab

Changed header from 'Changelog' to 'What's New'

kdk1616 committed:
Restricted to show only last two most recent versions (fixed formatting)

tektaxi reviewed (approved):
Looks good, glad to be able to see updates in Aspine now.

psvenk commented on public/home.html:
```diff
@@ -416,8 +416,20 @@ <h2 class="info-header">GPA Type and Switching Quarters</h2>
+			    font-size: 100%;
+			    text-decoration: none;
+		    }
+      </style>
```
This CSS rule should be moved to `home.css`.

psvenk commented on public/home.html:
```diff
@@ -416,8 +416,20 @@ <h2 class="info-header">GPA Type and Switching Quarters</h2>
+      <br>
+      <br>
+      <p class="info-text"><span id="updates"></span></p>
+      <center><a href="https://github.com/Aspine/aspine/blob/master/CHANGELOG.md">&lt&lt See Older Versions of Aspine &gt&gt</a>
```
- It seems that you have an unclosed `<center>` tag
- [The `<center>` tag is deprecated](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/center); you should assign an ID to the `<a>` and use CSS instead.

psvenk commented on public/js/home.js:
```diff
@@ -1264,6 +1264,31 @@ document.getElementById("default_open").click();
+
+
+
+
```
I don't think we should have this much whitespace. I would just remove all whitespace between the end of the `for` loop adn teh end of the callback function and leave only one line between the end of the entire AJAX call and the `//#endif` (for some reason GitHub isn't letting me put this as a "suggestion" with a diff).

psvenk commented on serve.js:
```diff
@@ -236,6 +245,11 @@ app.use((req, res) => {
+
+
+
+
```
As before, I would also remove this whitespace, especially because you did not touch this code and it would be good to have a cleaner overall diff.

psvenk commented on serve.js:
```diff
@@ -141,6 +143,13 @@ app.use('/fonts/material-icons/iconfont', express.static(
+
+app.get('/updates', async (req, res) => {
+    const changelog = await fs.promises.readFile(__dirname + '/CHANGELOG.md');
+    res.send(marked("" + changelog));
```
Instead of using type coercion (adding a string to `changelog` to convert it from a `Buffer` to a string), use [`Buffer.prototype.toString`](https://nodejs.org/api/buffer.html#buffer_buf_tostring_encoding_start_end):
```suggestion
    res.send(marked(changelog.toString()));
```

psvenk commented on public/js/home.js:
```diff
@@ -1264,6 +1264,31 @@ document.getElementById("default_open").click();
+$.ajax("/updates").then(upt => {
+	$("#updates").html(upt);
+	document.getElementById("changelog").outerHTML = "<h2 class='info-header'>Version History/What's New:</h2>";
+    var items = document.getElementById("updates").getElementsByTagName("h2");
```
Use `const` instead of `var` and [`document.querySelectorAll`](https://developer.mozilla.org/en-US/docs/Web/API/Document/querySelectorAll) instead of `getElementById` and `getElementsByTagName`:
```suggestion
    const items = document.querySelectorAll("#updates h2");
```

psvenk commented on public/js/home.js:
```diff
@@ -1264,6 +1264,31 @@ document.getElementById("default_open").click();
+    }
+
+
+});
```
The indentation for this entire callback should be normalized.

psvenk commented on public/js/home.js:
```diff
@@ -1264,6 +1264,31 @@ document.getElementById("default_open").click();
+    var items = document.getElementById("updates").getElementsByTagName("h2");
+	for (var i = 0; i < items.length; i++) {
+	    items[i].className = "info-header";
+	}
```
Instead of using a C-style `for` loop prone to off-by-one errors, this can be done more simply:
```suggestion
	items.forEach(x => { x.className = "info-header"; });
```

psvenk commented on public/js/home.js:
```diff
@@ -1264,6 +1264,31 @@ document.getElementById("default_open").click();
+    //Hide Everything but first and second versions
+    for(var i = 3; i < items.length; i ++) {
+        items[i].style = "display: none";
+    }
```
This can also be done more directly and in a way that avoids potential conflict with other styling (just good practice in general):
```suggestion
    items.slice(3).forEach(x => { x.style.setProperty("display", "none") });
```

psvenk commented on public/js/home.js:
```diff
@@ -1264,6 +1264,31 @@ document.getElementById("default_open").click();
+    items = document.getElementById("updates").getElementsByTagName("ul");
+    for(var i = 3; i < items.length; i ++) {
+        items[i].style = "display: none";
+    }
```
This can also be simplified:
```suggestion
    const subpoints = document.querySelectorAll("#updates ul");
    subpoints.slice(3).forEach(x => { x.style.setProperty("display", "none") });
```

psvenk reviewed (changes requested):
Looks good overall (from reading the diff), though I have not yet tested your code. Here are some code style changes (but take them with a grain of salt because I haven't tested any of my proposed changes).

notrodes commented:
Stop writing jquery!!

notrodes commented:
Im fine not rewriting stuff, just stop writing morehow hard is that!?!?!?

notrodes reviewed (dismissed):
delete $

kdk1616 committed:
Apply suggestions from code review

Co-authored-by: psvenk <45520974+psvenk@users.noreply.github.com>

kdk1616 committed:
CSS changes to changelog

Moved CSS to CSS file

kdk1616 committed:
Javascript changes to remove all but most recent two versions

kdk1616 committed:
Updated CSS margins and font sizes of changelog

psvenk commented on public/js/home.js:
```diff
@@ -1268,21 +1268,16 @@ $.ajax("/version").then(ver => $("#version").text(ver));
 	$("#updates").html(upt);
 	document.getElementById("changelog").outerHTML = "<h2 class='info-header'>Version History/What's New:</h2>";
-    const items = document.querySelectorAll("#updates h2");
+    var items = [...document.querySelectorAll("#updates h2")];
```
@kdk1616 Why did you change `const` to `var` here? We should always avoid `var` in favor of `const` or `let`.

psvenk commented on public/js/home.js:
```diff
@@ -1268,21 +1268,16 @@ $.ajax("/version").then(ver => $("#version").text(ver));
-    const items = document.querySelectorAll("#updates h2");
+    var items = [...document.querySelectorAll("#updates h2")];
 	items.forEach(x => { x.className = "info-header"; });
+    console.log(items);
```
Remove log statement
```suggestion
```

psvenk commented on public/js/home.js:
```diff
@@ -1268,21 +1268,16 @@ $.ajax("/version").then(ver => $("#version").text(ver));
-
-
+    const subpoints = [...document.querySelectorAll("#updates ul")];
+    console.log(subpoints);
```
Remove log statement
```suggestion
```

psvenk commented on public/css/home.css:
```diff
@@ -1256,3 +1256,25 @@ p#large_clock_period {
+#updates ul {
+  margin-bottom: 20px;
+}
+#moreVersions {
```
Use kebab-case for CSS IDs and classes, not camelCase
```suggestion
#more-versions {
```

psvenk commented on public/home.html:
```diff
@@ -416,8 +416,14 @@ <h2 class="info-header">GPA Type and Switching Quarters</h2>
+
+      <hr>
+      <p class="info-text" id="updates-container"><span id="updates"></span></p>
+      <p id="moreVersions"><a  href="https://github.com/Aspine/aspine/blob/master/CHANGELOG.md">&lt&lt See Older Versions of Aspine &gt&gt</a></p>
```
Remove extraneous space between `a` and `href` and prevent overlong lines; use
kebab-case instead of camelCase for CSS ID
```suggestion
      <p id="more-versions">
        <a href="https://github.com/Aspine/aspine/blob/master/CHANGELOG.md">
          &lt&lt See Older Versions of Aspine &gt&gt
        </a>
      </p>
```

kdk1616 committed:
Minor Changes to changelog formatting

Increased font size of "More Versions" link
Removed paragraph at top of versions list with "Sematic Versioning" text

psvenk committed:
Fix code style for version history display

psvenk committed:
Version history: verbiage and style changes

psvenk reviewed (approved):
Looks good to me; I think this is ready to merge pending @notrodes' approval. I noticed that the changelog inclusion does not work in Aspine Lite, but I chose to split that into a separate issue (#250) (a) this is a sufficiently small feature, (b) the link to GitHub still exists on Aspine Lite, and (c) implementing this would require considerable work on the build script.

psvenk committed:
Pre-render changelog on the server side

To improve performance, render the changelog when starting the server
instead of doing it whenever the endpoint is reached.
